src/NLog/Targets/DatabaseConnectionInfo.cs

```diff
@@ -0,0 +1,113 @@
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+// THE POSSIBILITY OF SUCH DAMAGE.
+// 
+
+#if !SILVERLIGHT && !__IOS__ && !__ANDROID__
+
+namespace NLog.Targets
+{
+    using System;
+    using System.Collections.Generic;
+    using System.ComponentModel;
+    using System.Data;
+    using System.Globalization;
+    using NLog.Config;
+    using NLog.Internal;
+    using NLog.Layouts;
+
+    /// <summary>
+    /// Information about database command + parameters.
+    /// </summary>
+    [NLogConfigurationItem]
+    public class DatabaseConnectionInfo
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DatabaseConnectionInfo"/> class.
+        /// </summary>
+        public DatabaseConnectionInfo()
+        {
+        }
+
+        /// <summary>
+        /// Gets or sets the DbConnection Property to update
+        /// </summary>
+        /// <docgen category='Connection Options' order='10' />
+        [RequiredParameter]
+        public string Name { get; set; }
+
+        /// <summary>
+        /// Gets or sets the value to assign on DbConnection
+        /// </summary>
+        /// <docgen category='Connection Options' order='10' />
+        [RequiredParameter]
+        public Layout Layout { get; set; }
+
+        /// <summary>
+        /// Gets or sets the type of the property
+        /// </summary>
+        /// <docgen category='Connection Options' order='10' />
+        [DefaultValue(typeof(string))]
+        public Type PropertyType { get; set; } = typeof(string);
+
+        /// <summary>
+        /// Gets or sets convert format of the property value .
+        /// </summary>
+        /// <docgen category='Parameter Options' order='8' />
+        [DefaultValue(null)]
+        public string Format { get; set; }
+
+        /// <summary>
+        /// Gets or sets the culture used for parsing property string-value for type-conversion
+        /// </summary>
+        /// <docgen category='Parameter Options' order='9' />
+        [DefaultValue(null)]
+        public CultureInfo Culture { get; set; }
+
+        internal bool SetPropertyValue(IDbConnection dbConnection, object propertyValue)
+        {
+            var dbConnectionType = dbConnection.GetType();
+            var propertySetterCache = _propertySetter;
+            if (propertySetterCache.Key.Key != Name || propertySetterCache.Key.Value != dbConnectionType)
+            {
+                var propertySetter = PropertySetter.GeneratePropertySetter(dbConnectionType, Name);
+                propertySetterCache = new KeyValuePair<KeyValuePair<string, Type>, IPropertySetter>(new KeyValuePair<string, Type>(Name, dbConnectionType), propertySetter);
```

Comment:
I really don't like to KeyValuePair (unreadable)... as we can't use c# 7 tuples, I propose:

- Creating a new type (class or struct) for ` KeyValuePair<KeyValuePair<string, Type>, IPropertySetter>`
- could be private nested
- Implement Equals for the check below.

e.g.

```c#
private class PropertySetterCacheItem
{
    [NotNull] public string Name { get; }
    [NotNull] public Type Type { get; }
    [NotNull] public IPropertySetter PropertySetter { get; }

    public PropertySetterCacheItem([NotNull] string name, [NotNull] Type type, [NotNull] IPropertySetter propertySetter)
    {
        Name = name ?? throw new ArgumentNullException(nameof(name));
        Type = type ?? throw new ArgumentNullException(nameof(type));
        PropertySetter = propertySetter ?? throw new ArgumentNullException(nameof(propertySetter));
    }


    /// <summary>
    /// Equals without IPropertySetter
    /// </summary>
    protected bool Equals(string name, Type type)
    {
        return string.Equals(Name, name, StringComparison.OrdinalIgnoreCase) && Type == type;
    }


}
```

Reply:
Sure just wanted to make it quick, since you were so eager on making a half solution. Now suddenly it has changed :)

Reply:
LOL I had the same with #3825 ...

Reply:
Copy paste my private class is also quick? ;)

Reply:
But now changed into struct.

Reply:
OK, I was under the impression that a ref type in a struct, makes a struct a bit useless.

Reply:
I like immutable structs. They are clean and doesn't allocate. And provide nearness in memory.